# ETL PROJECT - RETAIL TRANSACTIONS

The dataset it will be used in this project is from Kaggle plataform. It contains information about retail transactions conducted online. It contains information about customer purchases, including the invoice number, stock code, description of the items purchased, quantity, unit price, invoice date, customer ID, and country.


## 1.1 - Columns Dictionary



**1) InvoiceNo:** A unique identifier for each transaction or invoice.

**2) StockCode:** A code representing the stock or item purchased.

**3) Description:** A textual description of the item purchased.

**4) Quantity:** The quantity of the item purchased in each transaction.

**5) InvoiceDate:** The date and time when the transaction occurred.

**6) UnitPrice:** The price per unit of the item purchased.

**7) CustomerID:** The unique identifier for the customer making the purchase.

**8) Country:** The country where the transaction took place.


## 1- Extract


The table it will be used it was in Google Big Query plataform. So it will be necessary extract all the informations from there.

In [1]:
#import dependencies

# extract and transform

import os
import pandas as pd
from google.cloud import bigquery
from google.oauth2 import service_account
import pandas_gbq
import numpy as np

# load to postgreSQL

from sqlalchemy import create_engine
import psycopg2

In [2]:
path = 'C:/Users/vanes/OneDrive/Documentos/PROJETOS_PORTFOLIO/ETL/'

In [3]:

# Set the path to your service account key JSON file
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = path+"gbq.json"

# Your existing code
selectQuery = """SELECT * FROM etl-project-416319.project_1.retail_df"""
bigqueryClient = bigquery.Client()
df = bigqueryClient.query(selectQuery).to_dataframe()
df.to_csv("customer.csv", index=False)


In [4]:
df

InvoiceNo     StockCode                          Description  Quantity  \
0         537032         21275                                    ?       -30   
1         538090         20956                                    ?      -723   
2         539494         21479                                    ?       752   
3         540100         22837                                    ?      -106   
4         540558         21258                                    ?       -29   
...          ...           ...                                  ...       ...   
541904    554550        47566B  incorrectly credited C550456 see 47      1300   
541905    554553        47566B  incorrectly credited C550456 see 47     -1300   
541906    547704         21621  rcvd be air temp fix for dotcom sit       100   
541907    556955  gift_0001_20  to push order througha s stock was         10   
541908    562113         22016  Dotcomgiftshop Gift Voucher £100.00         1   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-03 16:50:00       0.00        nan  United Kingdom  
1      2010-12-09 14:48:00       0.00        nan  United Kingdom  
2      2010-12-20 10:36:00       0.00        nan  United Kingdom  
3      2011-01-04 16:53:00       0.00        nan  United Kingdom  
4      2011-01-10 10:04:00       0.00        nan  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-05-25 09:57:00       0.00        nan  United Kingdom  
541905 2011-05-25 09:59:00       0.00        nan  United Kingdom  
541906 2011-03-24 17:51:00       0.00        nan  United Kingdom  
541907 2011-06-16 09:04:00       0.00        nan  United Kingdom  
541908 2011-08-02 16:11:00      83.33        nan  United Kingdom  

[541909 rows x 8 columns]

## 2- Transform

In [5]:
df.shape

(541909, 8)

In [6]:
# checking for duplicates

df.duplicated().sum()

5268

In [7]:

# pick up duplicated datas
duplicated = df[df.duplicated()]


print(duplicated)

# Se desejar contar o número de duplicados encontrados
print("Número de duplicados:", len(duplicated))


       InvoiceNo StockCode                          Description  Quantity  \
2015      572344         M                               Manual        48   
2016      572344         M                               Manual        48   
2017      572344         M                               Manual        48   
2018      572344         M                               Manual        48   
2019      572344         M                               Manual        48   
...          ...       ...                                  ...       ...   
541306    558710    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER         1   
541308    563924    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER         1   
541309    563924    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER         1   
541326    577296    18097C  WHITE TALL PORCELAIN T-LIGHT HOLDER         1   
541481    574063     22577  WOODEN HEART CHRISTMAS SCANDINAVIAN         2   

               InvoiceDate  UnitPrice CustomerID         Country  
2015   2

In [8]:
# dropping duplicated data

df = df.drop_duplicates()
df

InvoiceNo     StockCode                          Description  Quantity  \
0         537032         21275                                    ?       -30   
1         538090         20956                                    ?      -723   
2         539494         21479                                    ?       752   
3         540100         22837                                    ?      -106   
4         540558         21258                                    ?       -29   
...          ...           ...                                  ...       ...   
541904    554550        47566B  incorrectly credited C550456 see 47      1300   
541905    554553        47566B  incorrectly credited C550456 see 47     -1300   
541906    547704         21621  rcvd be air temp fix for dotcom sit       100   
541907    556955  gift_0001_20  to push order througha s stock was         10   
541908    562113         22016  Dotcomgiftshop Gift Voucher £100.00         1   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-03 16:50:00       0.00        nan  United Kingdom  
1      2010-12-09 14:48:00       0.00        nan  United Kingdom  
2      2010-12-20 10:36:00       0.00        nan  United Kingdom  
3      2011-01-04 16:53:00       0.00        nan  United Kingdom  
4      2011-01-10 10:04:00       0.00        nan  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-05-25 09:57:00       0.00        nan  United Kingdom  
541905 2011-05-25 09:59:00       0.00        nan  United Kingdom  
541906 2011-03-24 17:51:00       0.00        nan  United Kingdom  
541907 2011-06-16 09:04:00       0.00        nan  United Kingdom  
541908 2011-08-02 16:11:00      83.33        nan  United Kingdom  

[536641 rows x 8 columns]

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 536641 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    536641 non-null  object        
 1   StockCode    536641 non-null  object        
 2   Description  536641 non-null  object        
 3   Quantity     536641 non-null  Int64         
 4   InvoiceDate  536641 non-null  datetime64[us]
 5   UnitPrice    536641 non-null  float64       
 6   CustomerID   536641 non-null  object        
 7   Country      536641 non-null  object        
dtypes: Int64(1), datetime64[us](1), float64(1), object(5)
memory usage: 37.4+ MB


In [10]:
# checking for columns types

df.dtypes

InvoiceNo              object
StockCode              object
Description            object
Quantity                Int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID             object
Country                object
dtype: object

It seems all columns are the correct type format. But there are some inconsistency in CustomerID values. From the information, there aren't any missing values, but when the dataframe was vizualized, there are some rows with 'nan' values. This indicates that those values are missing, but for some reason, the BigQuery understood them as a string. 

In [11]:
# checking for missing values

df.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [12]:
df.CustomerID.value_counts().sort_index()

CustomerID
12346.0         2
12347.0       182
12348.0        31
12349.0        73
12350.0        17
            ...  
18281.0         7
18282.0        13
18283.0       721
18287.0        70
nan        135037
Name: count, Length: 4373, dtype: int64

In [13]:
df[df['Description'] == 'nan']

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
57      536414     22139         nan        56 2010-12-01 11:52:00        0.0   
58      536545     21134         nan         1 2010-12-01 14:32:00        0.0   
59      536546     22145         nan         1 2010-12-01 14:33:00        0.0   
60      536547     37509         nan         1 2010-12-01 14:33:00        0.0   
61      536549    85226A         nan         1 2010-12-01 14:34:00        0.0   
...        ...       ...         ...       ...                 ...        ...   
1506    581199     84581         nan        -2 2011-12-07 18:26:00        0.0   
1507    581203     23406         nan        15 2011-12-07 18:31:00        0.0   
1508    581209     21620         nan         6 2011-12-07 18:35:00        0.0   
1509    581234     72817         nan        27 2011-12-08 10:33:00        0.0   
1510    581408     85175         nan        20 2011-12-08 14:06:00        0.0   

     CustomerID         Country  
57          nan  United Kingdom  
58          nan  United Kingdom  
59          nan  United Kingdom  
60          nan  United Kingdom  
61          nan  United Kingdom  
...         ...             ...  
1506        nan  United Kingdom  
1507        nan  United Kingdom  
1508        nan  United Kingdom  
1509        nan  United Kingdom  
1510        nan  United Kingdom  

[1454 rows x 8 columns]

The dataframe has missing values in the CustomerID column but they are interpreted as a string. If the remaining corresponding data for these missing IDs suggested that the purchase did not exist, one approach to consider would be to delete these records. Alternatively, if the objective of the analysis is to understand the customer profile and how they shop on the site, then these records without identification would hinder the analysis, and deleting them would be the best approach.

In this case, the way will be followed is to keep the NaN values for some reasons:

- 1: **Data Integrity Preservation:** By keeping NaN records, you are preserving the integrity of the original data. This means that you are not introducing fictitious or arbitrary data that could distort the analysis.

- 2: **Transparency and Replicability:** Maintaining NaN records allows for an accurate representation of the original data, making the ETL process more transparent and replicable. Other analysts can easily understand that missing values have been identified and have not been removed or altered arbitrarily.

- 3: **Analysis Flexibility**: Maintaining NaN records offers flexibility in how you handle these values during analysis. You can choose to exclude them from certain metrics or treat them differently depending on the specific requirements of the analysis.

- 4: **Better Tool Support:** Many data analysis tools, such as Python pandas libraries or data visualization tools, have built-in support for handling NaN values. This facilitates the processing and analysis of this data without the need for additional transformations.


In [14]:
# replace 'nan' values to NaN

df.loc[:, 'CustomerID'] = df['CustomerID'].replace('nan', np.nan)


In [15]:
df

InvoiceNo     StockCode                          Description  Quantity  \
0         537032         21275                                    ?       -30   
1         538090         20956                                    ?      -723   
2         539494         21479                                    ?       752   
3         540100         22837                                    ?      -106   
4         540558         21258                                    ?       -29   
...          ...           ...                                  ...       ...   
541904    554550        47566B  incorrectly credited C550456 see 47      1300   
541905    554553        47566B  incorrectly credited C550456 see 47     -1300   
541906    547704         21621  rcvd be air temp fix for dotcom sit       100   
541907    556955  gift_0001_20  to push order througha s stock was         10   
541908    562113         22016  Dotcomgiftshop Gift Voucher £100.00         1   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-03 16:50:00       0.00        NaN  United Kingdom  
1      2010-12-09 14:48:00       0.00        NaN  United Kingdom  
2      2010-12-20 10:36:00       0.00        NaN  United Kingdom  
3      2011-01-04 16:53:00       0.00        NaN  United Kingdom  
4      2011-01-10 10:04:00       0.00        NaN  United Kingdom  
...                    ...        ...        ...             ...  
541904 2011-05-25 09:57:00       0.00        NaN  United Kingdom  
541905 2011-05-25 09:59:00       0.00        NaN  United Kingdom  
541906 2011-03-24 17:51:00       0.00        NaN  United Kingdom  
541907 2011-06-16 09:04:00       0.00        NaN  United Kingdom  
541908 2011-08-02 16:11:00      83.33        NaN  United Kingdom  

[536641 rows x 8 columns]

In [16]:
# removing '.0' at the end of CustomerID values

df.loc[:, 'CustomerID'] = df['CustomerID'].str.rstrip('.0')

In [17]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64

In [18]:
df_null_customer=df[df['CustomerID'].isna()]

In [19]:
df_null_customer.describe()

Quantity                 InvoiceDate      UnitPrice
count   135037.0                      135037  135037.000000
mean    1.996868  2011-06-16 03:12:34.079104       8.078342
min      -9600.0         2010-12-01 11:52:00  -11062.060000
25%          1.0         2011-02-21 18:10:00       1.630000
50%          1.0         2011-06-27 15:27:00       3.290000
75%          3.0         2011-10-11 16:37:00       5.490000
max       5568.0         2011-12-09 10:26:00   17836.460000
std    66.705155                         NaN     151.924958

A second investigation to conduct in this dataset is to thoroughly examine the description values.

In [20]:
df['Description'].value_counts()

Description
WHITE HANGING HEART T-LIGHT HOLDER     2357
REGENCY CAKESTAND 3 TIER               2189
JUMBO BAG RED RETROSPOT                2156
PARTY BUNTING                          1720
LUNCH BAG RED RETROSPOT                1625
                                       ... 
BREAD BIN, DINER STYLE, MINT              1
CAKESTAND, 3 TIER, LOVEHEART              1
GREEN PENDANT SHELL NECKLACE              1
OPULENT VELVET SET/3 CANDLES              1
Dotcomgiftshop Gift Voucher £100.00       1
Name: count, Length: 4224, dtype: int64

In [21]:
df['Description'].sample(50)

125695                CARD PSYCHEDELIC APPLES
57302                     SPACEBOY LUNCH BOX 
406580        HOT WATER BOTTLE I AM SO POORLY
496137     FAIRY CAKE FLANNEL ASSORTED COLOUR
499356     MAKE YOUR OWN FLOWERPOWER CARD KIT
334907          GROW YOUR OWN PLANT IN A CAN 
271461            RIBBON REEL STRIPES DESIGN 
50454                     EASTER TIN KEEPSAKE
101127                 HAND WARMER UNION JACK
454620       SMALL IVORY HEART WALL ORGANISER
410270        NATURAL SLATE HEART CHALKBOARD 
16529                         JUMBO BAG PEARS
494298     CHARLIE+LOLA RED HOT WATER BOTTLE 
73249                    SMALL POPCORN HOLDER
182964              ALARM CLOCK BAKELIKE PINK
439889       PACK OF 12 PINK PAISLEY TISSUES 
329071          DOORMAT KEEP CALM AND COME IN
155329               CLASSIC GLASS COOKIE JAR
172321               REGENCY CAKESTAND 3 TIER
370772         KNEELING MAT HOUSEWORK  DESIGN
128644                EGG CUP MILKMAID INGRID
365612         FELTCRAFT PRINCESS 

It was found some missing values but as the same case to CustomerID column, these values are interpretad as strings. So, they are going to be replaced for NaN values.

In [22]:
df[df['Description']=='nan']

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
57      536414     22139         nan        56 2010-12-01 11:52:00        0.0   
58      536545     21134         nan         1 2010-12-01 14:32:00        0.0   
59      536546     22145         nan         1 2010-12-01 14:33:00        0.0   
60      536547     37509         nan         1 2010-12-01 14:33:00        0.0   
61      536549    85226A         nan         1 2010-12-01 14:34:00        0.0   
...        ...       ...         ...       ...                 ...        ...   
1506    581199     84581         nan        -2 2011-12-07 18:26:00        0.0   
1507    581203     23406         nan        15 2011-12-07 18:31:00        0.0   
1508    581209     21620         nan         6 2011-12-07 18:35:00        0.0   
1509    581234     72817         nan        27 2011-12-08 10:33:00        0.0   
1510    581408     85175         nan        20 2011-12-08 14:06:00        0.0   

     CustomerID         Country  
57          NaN  United Kingdom  
58          NaN  United Kingdom  
59          NaN  United Kingdom  
60          NaN  United Kingdom  
61          NaN  United Kingdom  
...         ...             ...  
1506        NaN  United Kingdom  
1507        NaN  United Kingdom  
1508        NaN  United Kingdom  
1509        NaN  United Kingdom  
1510        NaN  United Kingdom  

[1454 rows x 8 columns]

In [23]:
df.loc[:, 'Description'] = df['Description'].replace('nan', np.nan)

In [24]:
df['Description'].isna().sum()

1454

In [25]:
df[df['Description']=='?']

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
0     537032     21275           ?       -30 2010-12-03 16:50:00        0.0   
1     538090     20956           ?      -723 2010-12-09 14:48:00        0.0   
2     539494     21479           ?       752 2010-12-20 10:36:00        0.0   
3     540100     22837           ?      -106 2011-01-04 16:53:00        0.0   
4     540558     21258           ?       -29 2011-01-10 10:04:00        0.0   
5     540651     82599           ?      -290 2011-01-10 15:00:00        0.0   
6     541882     48185           ?       -57 2011-01-24 10:54:00        0.0   
7     542861    47591B           ?      -207 2011-02-01 11:47:00        0.0   
8     544410    84997C           ?      -741 2011-02-18 13:54:00        0.0   
9     544649     21621           ?      -400 2011-02-22 12:37:00        0.0   
10    546138     21815           ?      -300 2011-03-09 16:34:00        0.0   
11    546139     84988           ?      3000 2011-03-09 16:35:00        0.0   
12    546145     21537           ?       -80 2011-03-09 16:39:00        0.0   
13    546148    85049D           ?      -100 2011-03-09 16:43:00        0.0   
14    546149     20866           ?      -230 2011-03-09 17:02:00        0.0   
15    546245    85185B           ?      -259 2011-03-10 13:26:00        0.0   
16    547339    84406B           ?      -450 2011-03-22 11:51:00        0.0   
17    554852     21272           ?      -212 2011-05-27 09:24:00        0.0   
18    556231    85123A           ?      4000 2011-06-09 15:04:00        0.0   
19    556299     21937           ?      -161 2011-06-10 10:17:00        0.0   
20    557035     82583           ?      -220 2011-06-16 13:52:00        0.0   
21    557036     20867           ?      -271 2011-06-16 13:53:00        0.0   
22    558457     21147           ?       -95 2011-06-29 13:59:00        0.0   
23    559314     84584           ?       -50 2011-07-07 13:51:00        0.0   
24    559375     22051           ?      -300 2011-07-08 10:37:00        0.0   
25    559655     21166           ?      -124 2011-07-11 12:57:00        0.0   
26    559943     21578           ?      -251 2011-07-14 10:21:00        0.0   
27    561665     22171           ?       142 2011-07-28 16:55:00        0.0   
28    561927     22740           ?     -1060 2011-08-01 12:49:00        0.0   
29    562164     21257           ?       -94 2011-08-03 12:08:00        0.0   
30    562382     22927           ?       -40 2011-08-04 15:36:00        0.0   
31    562383     22523           ?       -90 2011-08-04 15:37:00        0.0   
32    562546     20694           ?      -110 2011-08-05 15:46:00        0.0   
33    565288     23135           ?       101 2011-09-02 11:43:00        0.0   
34    565297    84877D           ?      -203 2011-09-02 12:06:00        0.0   
35    565307     21830           ?      -443 2011-09-02 12:20:00        0.0   
36    567629     40016           ?      -752 2011-09-21 13:21:00        0.0   
37    567693    85199L           ?      -250 2011-09-21 17:26:00        0.0   
38    568302     20700           ?      -121 2011-09-26 12:46:00        0.0   
39    569855     23412           ?       -30 2011-10-06 14:27:00        0.0   
40    570110     22570           ?       -74 2011-10-07 12:27:00        0.0   
41    570112     22569           ?       -84 2011-10-07 12:29:00        0.0   
42    570712     22121           ?       -90 2011-10-12 10:15:00        0.0   
43    571024     22812           ?      -270 2011-10-13 12:18:00        0.0   
44    571116     23131           ?       -49 2011-10-13 17:46:00        0.0   
45    571126    72802A           ?       -50 2011-10-14 09:53:00        0.0   
46    572920    72803A           ?       117 2011-10-26 16:52:00        0.0   

   CustomerID         Country  
0         NaN  United Kingdom  
1         NaN  United Kingdom  
2         NaN  United Kingdom  
3         NaN  United Kingdom  
4         NaN  United Kingdom  
5         NaN  Uni

In [26]:
df[df['Description']=='??']

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
47    566449     20856          ??      -140 2011-09-12 16:07:00        0.0   
48    566712     22030          ??      -635 2011-09-14 12:04:00        0.0   
49    569886     90191          ??       -60 2011-10-06 15:24:00        0.0   
50    570599    84596G          ??      -365 2011-10-11 12:00:00        0.0   
51    572702    37479P          ??      -184 2011-10-25 14:33:00        0.0   
52    572749     22572          ??      -335 2011-10-25 16:52:00        0.0   
53    575615     82582          ??      -130 2011-11-10 12:51:00        0.0   

   CustomerID         Country  
47        NaN  United Kingdom  
48        NaN  United Kingdom  
49        NaN  United Kingdom  
50        NaN  United Kingdom  
51        NaN  United Kingdom  
52        NaN  United Kingdom  
53        NaN  United Kingdom

In [27]:
df[df['Description']=='???']

InvoiceNo StockCode Description  Quantity         InvoiceDate  UnitPrice  \
54    580547     21201         ???      -390 2011-12-05 09:29:00        0.0   

   CustomerID         Country  
54        NaN  United Kingdom

In the dataset, rows with Description values equal to '?', '??', or '???' exhibit consistent patterns:

- CustomerID fields are missing.
- UnitPrice is set to 0.00.

This pattern suggests that these entries may represent stock movements. An appropriate course of action would be to engage the relevant department within the company to ascertain the nature of these entries. For the purpose of this analysis, it will be assumed that these values represent stock movements. Accordingly:

- Negative quantities imply that specific products have been depleted from the stock.
- Positive quantities indicate that products have been added to the stock.

In [28]:
# replace '?', '??' and '???' values for 'stock movements'

df.loc[:, 'Description'] = df['Description'].replace(['?', '??', '???'], 'STOCK MOVEMENTS')

In [29]:
df[df['Description'].isin(['?', '??', '???'])]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

In [30]:
df_NaN_CustomerID = df[df['CustomerID'].isna() & ~df['Description'].isin(['?', '??', '???'])]

In [31]:
df_NaN_CustomerID['Description'].value_counts()

Description
DOTCOM POSTAGE                         693
JUMBO BAG RED RETROSPOT                497
JUMBO STORAGE BAG SUKI                 414
JUMBO SHOPPER VINTAGE RED PAISLEY      388
JUMBO BAG WOODLAND ANIMALS             372
                                      ... 
MOP PENDANT SHELL NECKLACE               1
MONEY BOX FIRST AID DESIGN               1
SET OF 3 CONEY ISLAND OVAL BOXES         1
on cargo order                           1
Dotcomgiftshop Gift Voucher £100.00      1
Name: count, Length: 3552, dtype: int64

In [32]:
df_NaN_CustomerID['Description'].fillna('', inplace=True)
df_NaN_CustomerID[df_NaN_CustomerID['Description'].str.islower()]

C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\801861334.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_NaN_CustomerID['Description'].fillna('', inplace=True)
C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\801861334.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_NaN_CustomerID['Description'].fillna('', inplace=True)


InvoiceNo     StockCode                          Description  Quantity  \
1511      580990         21109                                  wet       -48   
1512      561249      DCGS0073                                 ebay        -4   
1513      561252      DCGS0069                                 ebay        -5   
1514      561253      DCGS0068                                 ebay       -10   
1515      561254      DCGS0067                                 ebay       -11   
...          ...           ...                                  ...       ...   
541901    566327         23084  allocate stock for dotcom orders ta         4   
541902    550460        47556B  did  a credit  and did not tick ret      1300   
541903    569341         82600  historic computer difference?....se      -458   
541906    547704         21621  rcvd be air temp fix for dotcom sit       100   
541907    556955  gift_0001_20  to push order througha s stock was         10   

               InvoiceDate  UnitPrice CustomerID         Country  
1511   2011-12-06 16:54:00        0.0        NaN  United Kingdom  
1512   2011-07-26 11:51:00        0.0        NaN  United Kingdom  
1513   2011-07-26 11:52:00        0.0        NaN  United Kingdom  
1514   2011-07-26 11:52:00        0.0        NaN  United Kingdom  
1515   2011-07-26 11:52:00        0.0        NaN  United Kingdom  
...                    ...        ...        ...             ...  
541901 2011-09-12 12:13:00        0.0        NaN  United Kingdom  
541902 2011-04-18 13:18:00        0.0        NaN  United Kingdom  
541903 2011-10-03 14:19:00        0.0        NaN  United Kingdom  
541906 2011-03-24 17:51:00        0.0        NaN  United Kingdom  
541907 2011-06-16 09:04:00        0.0        NaN  United Kingdom  

[493 rows x 8 columns]

There are some records in the original dataframe with descriptions in lowercase. This operation will replace them, transforming all descriptions into uppercase.

In [33]:
# replace lowercase into uppercase for 'Description' column.

df.loc[:, 'Description'] = df['Description'].str.upper()

In [34]:
# replace lowercase into uppercase for 'Country' column.

df.loc[:, 'Country'] = df['Country'].str.upper()

In [35]:
# replace lowercase into uppercase for 'StockCode' column.

df.loc[:, 'StockCode'] = df['StockCode'].str.upper()

In [36]:
# replace lowercase into uppercase for 'InvoiceNo' column.

df.loc[:, 'InvoiceNo'] = df['InvoiceNo'].str.upper()

In [37]:
df

InvoiceNo     StockCode                          Description  Quantity  \
0         537032         21275                      STOCK MOVEMENTS       -30   
1         538090         20956                      STOCK MOVEMENTS      -723   
2         539494         21479                      STOCK MOVEMENTS       752   
3         540100         22837                      STOCK MOVEMENTS      -106   
4         540558         21258                      STOCK MOVEMENTS       -29   
...          ...           ...                                  ...       ...   
541904    554550        47566B  INCORRECTLY CREDITED C550456 SEE 47      1300   
541905    554553        47566B  INCORRECTLY CREDITED C550456 SEE 47     -1300   
541906    547704         21621  RCVD BE AIR TEMP FIX FOR DOTCOM SIT       100   
541907    556955  GIFT_0001_20  TO PUSH ORDER THROUGHA S STOCK WAS         10   
541908    562113         22016  DOTCOMGIFTSHOP GIFT VOUCHER £100.00         1   

               InvoiceDate  UnitPrice CustomerID         Country  
0      2010-12-03 16:50:00       0.00        NaN  UNITED KINGDOM  
1      2010-12-09 14:48:00       0.00        NaN  UNITED KINGDOM  
2      2010-12-20 10:36:00       0.00        NaN  UNITED KINGDOM  
3      2011-01-04 16:53:00       0.00        NaN  UNITED KINGDOM  
4      2011-01-10 10:04:00       0.00        NaN  UNITED KINGDOM  
...                    ...        ...        ...             ...  
541904 2011-05-25 09:57:00       0.00        NaN  UNITED KINGDOM  
541905 2011-05-25 09:59:00       0.00        NaN  UNITED KINGDOM  
541906 2011-03-24 17:51:00       0.00        NaN  UNITED KINGDOM  
541907 2011-06-16 09:04:00       0.00        NaN  UNITED KINGDOM  
541908 2011-08-02 16:11:00      83.33        NaN  UNITED KINGDOM  

[536641 rows x 8 columns]

In [38]:

# Converter 'InvoiceDate' para datetime, caso ainda não esteja
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])

# Extrair apenas o mês e o ano e armazenar em uma nova coluna chamada 'MonthYear'
df.loc[:, 'MonthYear'] = df['InvoiceDate'].dt.to_period('M')

# Converter 'MonthYear' para timestamp
df.loc[:, 'MonthYear'] = df['MonthYear'].dt.to_timestamp()



C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\2249233402.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\2249233402.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[:, 'MonthYear'] = df['InvoiceDate'].dt.to_period('M')
C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\2249233402.py:8: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pand

In [39]:
df

InvoiceNo     StockCode                          Description  Quantity  \
0         537032         21275                      STOCK MOVEMENTS       -30   
1         538090         20956                      STOCK MOVEMENTS      -723   
2         539494         21479                      STOCK MOVEMENTS       752   
3         540100         22837                      STOCK MOVEMENTS      -106   
4         540558         21258                      STOCK MOVEMENTS       -29   
...          ...           ...                                  ...       ...   
541904    554550        47566B  INCORRECTLY CREDITED C550456 SEE 47      1300   
541905    554553        47566B  INCORRECTLY CREDITED C550456 SEE 47     -1300   
541906    547704         21621  RCVD BE AIR TEMP FIX FOR DOTCOM SIT       100   
541907    556955  GIFT_0001_20  TO PUSH ORDER THROUGHA S STOCK WAS         10   
541908    562113         22016  DOTCOMGIFTSHOP GIFT VOUCHER £100.00         1   

               InvoiceDate  UnitPrice CustomerID         Country  MonthYear  
0      2010-12-03 16:50:00       0.00        NaN  UNITED KINGDOM 2010-12-01  
1      2010-12-09 14:48:00       0.00        NaN  UNITED KINGDOM 2010-12-01  
2      2010-12-20 10:36:00       0.00        NaN  UNITED KINGDOM 2010-12-01  
3      2011-01-04 16:53:00       0.00        NaN  UNITED KINGDOM 2011-01-01  
4      2011-01-10 10:04:00       0.00        NaN  UNITED KINGDOM 2011-01-01  
...                    ...        ...        ...             ...        ...  
541904 2011-05-25 09:57:00       0.00        NaN  UNITED KINGDOM 2011-05-01  
541905 2011-05-25 09:59:00       0.00        NaN  UNITED KINGDOM 2011-05-01  
541906 2011-03-24 17:51:00       0.00        NaN  UNITED KINGDOM 2011-03-01  
541907 2011-06-16 09:04:00       0.00        NaN  UNITED KINGDOM 2011-06-01  
541908 2011-08-02 16:11:00      83.33        NaN  UNITED KINGDOM 2011-08-01  

[536641 rows x 9 columns]

In [40]:
df.rename(columns={
    'InvoiceNo': 'invoice_no',
    'StockCode': 'stock_code',
    'Description': 'description',
    'Quantity': 'quantity',
    'InvoiceDate': 'invoice_date',
    'UnitPrice': 'unit_price',
    'CustomerID': 'customer_id',
    'Country': 'country',
    'MonthYear': 'month_year'
}, inplace=True)


C:\Users\vanes\AppData\Local\Temp\ipykernel_5552\3045793805.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.rename(columns={


In [41]:
df.columns

Index(['invoice_no', 'stock_code', 'description', 'quantity', 'invoice_date',
       'unit_price', 'customer_id', 'country', 'month_year'],
      dtype='object')

With the table cleaned and values handled, the dataframe will be imported into a PostgreSQL database. 

## 3-  Load to PostgreSQL

This part of project will be divided in some steps. First of all, it will be created a connection between Pandas and PostgreSQL. The follow stes is create a database that it will be called 'project_etl', after the dataframe pandas will be insert in this database.

### 3.1 - Creating a connection

In [42]:
# creating connection to postgreSQL

pgconn = psycopg2.connect(
    host='localhost',
    user = 'postgres',
    password='admin',
    database = 'postgres'

)

In [43]:
# cursor

pgcursor = pgconn.cursor()

In [44]:
# required code

from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
pgconn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)

In [45]:
# disconnect all connection were activated

pgcursor.execute("""
SELECT pg_terminate_backend(pg_stat_activity.pid)
FROM pg_stat_activity
WHERE pg_stat_activity.datname = 'db' -- Replace 'db' with your database name
  AND pid <> pg_backend_pid();
""")

pgconn.commit()


### 3.2 - Create database and disconnect from it

In [46]:
# verify if the database exists
pgcursor.execute("SELECT 1 FROM pg_database WHERE datname = 'project_etl'")
exists = pgcursor.fetchone()

# if doesn't exist, it will be created
if not exists:
    pgcursor.execute('CREATE DATABASE project_etl')


In [47]:
# commit (to salve all changes)

pgconn.commit()

In [48]:
# close
pgconn.close()

### 3.3- Insert the dataframe into  the database was created

In [49]:
pgconn = psycopg2.connect(
    host='localhost',
    user = 'postgres',
    password='admin',
    database = 'project_etl'

)

In [50]:
# cursor

pgcursor = pgconn.cursor()

In [51]:
# required code

from psycopg2.extensions import ISOLATION_LEVEL_AUTOCOMMIT
pgconn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)

In [52]:
# checking which db is being used.

pgcursor.execute('SELECT current_database()')

pgcursor.fetchone()

('project_etl',)

In [53]:
# create engine

# connection string: dialect+driver://username:password@server/database

variable= 'postgresql+psycopg2://postgres:admin@localhost/project_etl'
engine = create_engine(variable)

In [54]:
#display engine info

engine

Engine(postgresql+psycopg2://postgres:***@localhost/project_etl)

In [55]:
df.to_sql('orders_historical', engine, if_exists='replace', index=False)

641

for chunk in pd.read_csv('customer.csv',chunksize=536641):
    chunk.to_sql('orders', engine, if_exists ='replace')